File: ScrubAndExportProject.rb
#@category 0Saverio/DosMalware

java_import 'ghidra.app.util.exporter.AsciiExporter'
java_import 'ghidra.app.util.exporter.XmlExporter'
java_import 'ghidra.app.util.Option'
java_import 'ghidra.util.task.TaskMonitor'
java_import 'java.io.File'

project_dir = currentProgram.getDomainFile.getProjectLocator.getLocation

# Export ASM listing

asm_path = "#{project_dir}#{currentProgram.getName}.asm"

asm_exporter = AsciiExporter.new
asm_exporter.get_options { currentProgram }
asm_exporter.set_options([
  Option.new('Field Widths', ' Address ',     java.lang.Integer.new(25)),
  Option.new('Field Widths', ' Operand ',     java.lang.Integer.new(60)),
  Option.new('Field Widths', ' End of Line ', java.lang.Integer.new(80)),
])
asm_exporter.export(File.new(asm_path), currentProgram, nil, TaskMonitor::DUMMY)

processed_asm = IO
  .read(asm_path)
  .gsub(/\s+$/, '')

IO.write(asm_path, processed_asm)

puts "Listing exported to: #{asm_path}"

# Export XML and scrub it

xml_path = "#{project_dir}#{currentProgram.getName}.xml"

XmlExporter.new.export(File.new(xml_path), currentProgram, nil, TaskMonitor::DUMMY)

# Watch out the `\S` - `.` would eat the closing tag.
processed_xml = IO
  .read(xml_path)
  .gsub(%r{#{Dir.home}\S*/}, '/path/to/')
  .gsub(Etc.getlogin, 'myuser')
  .gsub(/TIMESTAMP=".+?\K\d\d:\d\d:\d\d/, '00:00:00')

IO.write(xml_path, processed_xml)

puts "XML exported to: #{xml_path}"

File: ScrubAndExportProject.java
// Exports ASM listing and XML, then scrubs personal info from both.
//
// @category 0Saverio/General

import ghidra.app.script.GhidraScript;
import ghidra.app.util.exporter.AsciiExporter;
import ghidra.app.util.exporter.XmlExporter;
import ghidra.app.util.Option;
import java.io.File;
import java.nio.file.*;
import java.util.List;
import java.util.regex.Pattern;

public class ScrubAndExportProject extends GhidraScript {

  @Override
  protected void run() throws Exception {
    String projectDir = currentProgram.getDomainFile()
        .getProjectLocator().getLocation().toString();
    if (!projectDir.endsWith(File.separator)) {
      projectDir += File.separator;
    }
    String progName = currentProgram.getName();

    // Export ASM listing

    String asmPath = projectDir + progName + ".asm";

    AsciiExporter asmExporter = new AsciiExporter();
    List<Option> opts = asmExporter.getOptions(() -> currentProgram);
    for (Option opt : opts) {
      String name = opt.getName().trim();
      if (name.equals("Address"))
        opt.setValue(Integer.valueOf(25));
      else if (name.equals("Operand"))
        opt.setValue(Integer.valueOf(60));
      else if (name.equals("End of Line"))
        opt.setValue(Integer.valueOf(80));
    }
    asmExporter.setOptions(opts);
    asmExporter.export(new File(asmPath), currentProgram, null, monitor);

    String asmContent = Files.readString(Path.of(asmPath));
    asmContent = asmContent.replaceAll("(?m)\\s+$", "");
    Files.writeString(Path.of(asmPath), asmContent);

    printf("Listing exported to: %s\n", asmPath);

    // Export XML and scrub it

    String xmlPath = projectDir + progName + ".xml";

    new XmlExporter().export(new File(xmlPath), currentProgram, null, monitor);

    String homeDir = System.getProperty("user.home");
    String username = System.getProperty("user.name");
    String xmlContent = Files.readString(Path.of(xmlPath));
    xmlContent = xmlContent.replaceAll(Pattern.quote(homeDir) + "\\S*/", "/path/to/");
    xmlContent = xmlContent.replace(username, "myuser");
    xmlContent = xmlContent.replaceAll("(TIMESTAMP=\"[^\"]*?)\\d\\d:\\d\\d:\\d\\d", "$100:00:00");
    Files.writeString(Path.of(xmlPath), xmlContent);

    printf("XML exported to: %s\n", xmlPath);
  }
}
